snap/quota/quota.go

```diff
@@ -182,22 +251,76 @@ func (grp *Group) validate() error {
 		}
 	}
 
-	// check that if this is a sub-group, then the parent group has enough space
+	// Check that if this is a sub-group, then the parent group has enough space
 	// to accommodate this new group (we assume that other existing sub-groups
-	// in the parent group have already been validated)
+	// in the parent group have already been validated).
 	if grp.parentGroup != nil {
-		alreadyUsed := quantity.Size(0)
+		memoryUsed := quantity.Size(0)
+		tasksUsed := 0
+		cpuQuotaUsed := 0
+
+		max := func(a, b int) int {
+			if a > b {
+				return a
+			}
+			return b
+		}
+
+		calculateCpuQuota := func(grp *Group) int {
+			if grp.CpuLimit == nil {
+				return 0
+			}
+			return max(grp.CpuLimit.Count, 1) * grp.CpuLimit.Percentage
+		}
+
 		for _, child := range grp.parentGroup.subGroups {
 			if child.Name == grp.Name {
 				continue
 			}
-			alreadyUsed += child.MemoryLimit
+
+			memoryUsed += child.MemoryLimit
+			tasksUsed += child.TaskLimit
+			cpuQuotaUsed += calculateCpuQuota(child)
+		}
+
+		if grp.parentGroup.CpuLimit != nil && grp.CpuLimit != nil {
+			arrayContains := func(arr []int, val int) bool {
+				for _, v := range arr {
+					if v == val {
+						return true
+					}
+				}
+				return false
+			}
+
+			upperLimit := calculateCpuQuota(grp.parentGroup)
+			if cpuQuotaUsed+calculateCpuQuota(grp) > upperLimit {
+				return fmt.Errorf("group %q would exceed its parent group's CPU limit", grp.Name)
+			}
+
+			if len(grp.parentGroup.CpuLimit.AllowedCPUs) > 0 {
+				// check if the group's allowed cpus are a subset of the parent group's allowed cpus
+				for _, cpu := range grp.CpuLimit.AllowedCPUs {
+					if !arrayContains(grp.parentGroup.CpuLimit.AllowedCPUs, cpu) {
+						return fmt.Errorf("group %q has a CPU allowance that is not allowed by its parent group", grp.Name)
+					}
+				}
+			}
+		}
+
+		if grp.parentGroup.MemoryLimit != 0 {
+			// careful arithmetic here in case we somehow overflow the max size of
+			// quantity.Size
+			if grp.MemoryLimit+memoryUsed > grp.parentGroup.MemoryLimit {
+				remaining := grp.parentGroup.MemoryLimit - memoryUsed
+				return fmt.Errorf("sub-group memory limit of %s is too large to fit inside remaining quota space %s for parent group %s", grp.MemoryLimit.IECString(), remaining.IECString(), grp.parentGroup.Name)
+			}
 		}
-		// careful arithmetic here in case we somehow overflow the max size of
-		// quantity.Size
-		if grp.parentGroup.MemoryLimit-alreadyUsed < grp.MemoryLimit {
-			remaining := grp.parentGroup.MemoryLimit - alreadyUsed
-			return fmt.Errorf("sub-group memory limit of %s is too large to fit inside remaining quota space %s for parent group %s", grp.MemoryLimit.IECString(), remaining.IECString(), grp.parentGroup.Name)
+
+		if grp.parentGroup.TaskLimit != 0 {
```

Comment:
shouldn't we also do the same check we do for memory for TaskLimit here to ensure that this new child task limit fits inside the parent, while including all the other sibling groups' TaskLimits ?

Reply:
Really really good point - this case I had not accounted for at all. I am rewriting the entire code that verifies whether a limit fits inside a group that takes into account all parent and sub-groups.